lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+  # inherited attributes will be included in the returned `Hash`.
+  #
+  # @param include_parent [Boolean] `true` if inherited attributes should be included
+  # @return [Hash{String=>PAttribute}] a hash with the attributes
+  # @api public
+  def attributes(include_parent = false)
+    get_members(include_parent, :attributes)
+  end
+
+  # Returns the attributes that participate in equality comparison. Inherited equality attributes
+  # are included.
+  # @return [Hash{String=>PAttribute}] a hash of attributes
+  # @api public
+  def equality_attributes
+    all = {}
+    collect_equality_attributes(all)
+    all
+  end
+
+  # @return [Boolean] `true` if this type is included when comparing instances
+  # @api public
+  def equality_include_type?
+    @equality_include_type
+  end
+
+  # Returns the functions of this `Object` type. If _include_parent_ is `true`, then all
+  # inherited functions will be included in the returned `Hash`.
+  #
+  # @param include_parent [Boolean] `true` if inherited functions should be included
+  # @return [Hash{String=>PFunction}] a hash with the functions
+  # @api public
+  def functions(include_parent = false)
+    get_members(include_parent, :functions)
+  end
+
+  DEFAULT = PObjectType.new(EMPTY_HASH)
+  # Assert that this type does not inherit from itself
+  # @api private
+  def check_self_recursion(originator)
+    unless @parent.nil?
+      raise Puppet::Error, "The Object type aliased by '#{originator}' inherits from itself" if @parent.equal?(originator)
+      @parent.check_self_recursion(originator)
+    end
+  end
+
+  protected
+
+  # An Object type is only assignable from another Object type. The other type
+  # or one of its parents must be equal to this type.
+  def _assignable?(o, guard)
+    if self == o
+      true
+    else
+      if o.is_a?(PObjectType)
+        op = o.parent
+        op.nil? ? false : assignable?(op, guard)
+      else
+        false
+      end
+    end
+  end
+
+  def get_members(include_parent, member_type)
+    all = {}
+    collect_members(all, include_parent, member_type)
+    all
+  end
+
+  def collect_members(collector, include_parent, member_type)
+    if include_parent
+      parent = resolved_parent
+      parent.collect_members(collector, include_parent, member_type) if parent.is_a?(PObjectType)
+    end
+    collector.merge!(@attributes) unless member_type == :functions
+    collector.merge!(@functions) unless member_type == :attributes
+    nil
+  end
+
+  def collect_equality_attributes(collector)
+    parent = resolved_parent
+    parent.collect_equality_attributes(collector) if parent.is_a?(PObjectType)
+    if @equality.nil?
+      # All attributes except constants participate
+      collector.merge!(@attributes.reject { |_, attr| attr.kind == ATTRIBUTE_KIND_CONSTANT })
+    else
+      collector.merge!(Hash[@equality.map { |attr_name| [attr_name, @attributes[attr_name]] }])
+    end
+    nil
+  end
+
+  # @api private
+  def label
+    'object'
```

Comment:
Sure, it could be frozen. But in general I try to freeze things where it matters (for performance or because immutability must be protected). This label is @api private and only used for errors.